### ISSUE TITLE
Addressing Input buffer & Ping

### ISSUE THREAD

**Issue by Uka-Uka** [brunibunny/pcsx2-online#1]
Hey, my man! I'm really glad to see you picking up the project after SmileTheory, I was desperately looking for that memcard synchronization! and I have some suggestions if you allow ofc...

- I believe that Alexsharoff's build was applying the buffering on both the host & client, then Smile just decided to modify that aspect to have that delay applied to the client only, it'd be really nice if you can reverse this again unless it was implemented for a reason. maybe an option to toggle between these two methods? that's too much to ask sorry :P

- Also, a ping indicator in the chat lobby would be a really nice addition, BTW, How should we adjust the input delay value properly? we used to just divide ping by 15 just like Dolphin's netplay guide says.

**Comment by brunibunny** [comment]
yo! sorry for taking a while to answer. I've just released a test version with the changes you requested. Let me know if any issues show up.
The ping indicator could use some more polish, but it will do for now.
I wouldn't say I'm experienced enough to pickup the project after SmileTheory or alexsharoff... but I'm trying to at least make things a bit more confortable to players. This is a learning experience for me as well :p

**Comment by nipkownix** [comment]
yo! sorry for taking a while to answer. I've just released a test version with the changes you requested. Let me know if any issues show up.
The ping indicator could use some more polish, but it will do for now.
I wouldn't say I'm experienced enough to pickup the project after SmileTheory or alexsharoff... but I'm trying to at least make things a bit more confortable to players. This is a learning experience for me as well.

About how to calculate the input delay value properly... I'm not too sure about that. I'd say the folks who worte Dolphin's netplay guide are more experienced than me :p

**Comment by Uka-Uka** [comment]
Not at all, bro! you've delivered the changes sooner than I could ever imagine.

I believe that applying the delay to all the players shouldn't be a luxury option, it's a must, it's a game-changer, the gameplay went so smooth and balanced, I and my friend couldn't even notice the input lag, unlike the previous versions it was a pain on the client-side.
Also, it could be a placebo but it worth mentioning that we've played nearly 3hrs without any desyncs, I think we hadn't such a steady session like that one before.

As for the ping indicator, it's doing the job and we're really thankful for the addition but as you've said it needs some refinements. it just acts strangely after starting the game.

One last thing related to the input buffer, Alexsharoff's build has an automatic input delay detection, it gives you a ballpark value based on ping which you still can override it if you want but that feature got removed by Smile, again I don't know why but maybe you can figure out the formula from Alex code and reimplement that feature again.

Anyway man, thanks a million! we really appreciate your time & efforts and wish you a useful learning journey.

**Comment by Uka-Uka** [comment]
Update: I've asked Alexsharoff about how his build is calculating the input delay, he has said "(average ping between players / 33,333)"

**Comment by Uka-Uka** [comment]
Update: There's really no need to reimplement that feature as it seems to be so problematic, it's a must to adjust that value manually especially with more than 2 players, it just would be worthy to add a hint in the lobby dialogue that a good starting value would be (Highest ping among players / 30)

Feel free to close this issue if you're done with it.

**Comment by nipkownix** [comment]
> There's really no need to reimplement that feature as it seems to be so problematic, it's a must to adjust that value manually especially with more than 2 players

Why is it problematic, exactly? I think it might still be worth a shot, right?

**Comment by Uka-Uka** [comment]
> 
> 
> > There's really no need to reimplement that feature as it seems to be so problematic, it's a must to adjust that value manually especially with more than 2 players
> 
> Why is it problematic, exactly? I think it might still be worth a shot, right?

Yeah, ofc players can make use of it, but I meant it's problematic programming-wise, I'm not a programmer so I shouldn't have said that but I can imagine that to get things right, the code needs to repeat and update itself every time a new player joins the lobby to consider his ping as you need to use the highest ping in there as a baseline, that comes in handy when playing with more than 2 players.

So the thing is that I didn't want to weigh you down and I thought that providing a hint about how to calculate the input delay would be easier, but yeah bringing that feature back would be cool, so it's your call now :v:

**Comment by Uka-Uka** [comment]
Another thing to keep in mind that the input delay calculating formula is assuming every player is able to run the game full speed at 30fps hence we divide the ping/33,333 but if there a player with a low-end pc running the game at let's say 24fps, his frame time will be more than 33,333ms so again this feature would only give a good starting value that needs to be adjusted manually if needed in most cases.

**Comment by BevP90** [comment]
> Hey, my man! I'm really glad to see you picking up the project after SmileTheory, I was desperately looking for that memcard synchronization! and I have some suggestions if you allow ofc...
> 
> * I believe that Alexsharoff's build was applying the buffering on both the host & client, then Smile just decided to modify that aspect to have that delay applied to the client only, it'd be really nice if you can reverse this again unless it was implemented for a reason. maybe an option to toggle between these two methods? that's too much to ask sorry :P
> * Also, a ping indicator in the chat lobby would be a really nice addition, BTW, How should we adjust the input delay value properly? we used to just divide ping by 15 just like Dolphin's netplay guide says.

@brunibunny Brilliant work. Thank you very much for picking this back up. Your improvements are huge.
@Uka-Uka I'd hope not. Older builds applying input delay to the client side was a huge issue for our Discord and made it unplayable and also given host advantage, it just wasn't right or fair for tournaments.
What I'm saying is we should have the option as we do now, I don't think it should change. Keep it fair.

**Comment by BevP90** [comment]
@brunibunny My question to you is this. Given that the input delay feels quite heavy, is there any way we can make inputs more snappy over netplay to make it more playable? We're finding delays 4 and 5 a bit chunky! I do however appreciate this is a lot of work and that you are trying, so thank you.
Back on the Dolphin front, that whole input buffer system is light years ahead, I couldn't believe that it was delay netcode it was so good, put the buffer up to 90 and it still felt great.
